File: 2024/python/day_7.py
import aoc

data = aoc.import_input(7)

total = 0
total2 = 0


for line in data.split("\n"):
    nums = aoc.get_nums(line)
    target, *operands = nums
    for i in range(2**len(operands)):
        cur = operands[0]
        for j in range(1, len(operands)):
            if i & (1 << j):
                cur += operands[j]
            else:
                cur *= operands[j]
        
        if cur > target:
            continue

        if cur == target:
            total += target
            break
    
    for i in range(3**len(operands)):
        cur = operands[0]
        for j in range(1, len(operands)):
            if i % 3 == 0:
                cur += operands[j]
            elif i % 3 == 1:
                cur *= operands[j]
            else:
                cur = int(str(cur) + str(operands[j]))
            i //= 3

        if cur > target:
            continue
            
        if cur == target:
            total2 += target
            break


print(total)
print(total2)

File: 2024/java/Day7.java

import aoc.FileHandler;
import java.util.stream.Stream;

public class Day7 {
    public static void main(String[] args) {
        String input = FileHandler.getInput(7);
        String[] lines = input.split("\n");
        long count1 = 0;
        long count2 = 0;
        for (String line : lines) {
            long target = parseTarget(line);
            int[] nums = parseNumbers(line);
            if (part1(target, nums)) {
                count1 += target;
            }
            if (part2(target, nums)) {
                count2 += target;
            }
        }
        System.out.println(count1);
        System.out.println(count2);
    }

    private static long parseTarget(String line) {
        return Long.parseLong(line.split(":")[0]);
    }

    private static int[] parseNumbers(String line) {
        return Stream.of(line.split(": ")[1].split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
    }

    public static boolean part1(long target, int[] nums) {
        for (int i = 0; i < Math.pow(2, nums.length - 1); i++) {
            int ops = i;
            int j = 1;
            long total = nums[0];
            for (int k = 0; k < nums.length - 1; k++) {
                if ((ops % 2) == 0) {
                    total += nums[j];
                } else {
                    total *= nums[j];
                }
                ops /= 2;
                j++;
            }
            if (total == target) {
                return true;
            }
        }
        return false;
    }

    public static boolean part2(long target, int[] nums) {
        for (int i = 0; i < Math.pow(3, nums.length - 1); i++) {
            int ops = i;
            int j = 1;
            long total = nums[0];
            for (int k = 0; k < nums.length - 1; k++) {
                if ((ops % 3) == 0) {
                    total += nums[j];
                } else if ((ops % 3) == 1) {
                    total *= nums[j];
                } else {
                    total = Long.parseLong("" + total + nums[j]);
                }
                ops /= 3;
                j++;
            }
            if (total == target) {
                return true;
            }
        }
        return false;
    }

}
